### PR TITLE
Smart Contracts playbooks rework

### LINKED ISSUE #929: Improve playbooks with smart contract examples
Due to the new changes in the smart contract language, we should add few non-regression tests for smart contract deployment and interaction, to make code proposal safer.

---

Epic: #154

### PR DESCRIPTION
**Pull request by bchamagne**
# Description

Freshen up the smart contract playbooks. 
They are easier to write, easier to read and the logs are easier on the eyes. 
Add 3 new test contracts: 

1. Faucet
1. Counter
1. UCO All-Time High

Depends on  #995 (Json.parse function)
Fixes #929

## Type of change

- Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

Start a local node and then run `mix archethic.regression --playbook localhost`.

# Checklist:

- My code follows the style guidelines of this project
- I have performed a self-review of my own code
- I have commented my code, particularly in hard-to-understand areas
- I have made corresponding changes to the documentation
- My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature works
- New and existing unit tests pass locally with my changes
- Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: lib/archethic/contracts/worker.ex [hunk 1]
```diff
@@ -196,27 +200,52 @@ defmodule Archethic.Contracts.Worker do
...
+    # because one of them would get a "transaction is already mining"
+    next_tick =
+      if {:datetime, next_tick} in triggers_type do
+        Logger.debug(
+          "Contract scheduler skips next tick for trigger=interval because there is a trigger=datetime at the same time that takes precedence"
+        )
+
+        interval
+        |> CronParser.parse!(@extended_mode?)
+        |> CronScheduler.get_next_run_date!(next_tick |> DateTime.add(1) |> DateTime.to_naive())
```
Comment (Neylix) [inline lib/archethic/contracts/worker.ex:223]: You can use
```elixir
Utils.next_date(interval, next_tick)
```

File: lib/archethic/contracts/worker.ex [hunk 2]
```diff
@@ -196,27 +200,52 @@ defmodule Archethic.Contracts.Worker do
     {:via, Registry, {ContractRegistry, address}}
   end
 
-  defp schedule_trigger(trigger = {:interval, interval}) do
+  defp schedule_trigger(trigger = {:interval, interval}, triggers_type) do
+    now = DateTime.utc_now()
+
+    # same code as Utils.time_offset/3 but we need the intermediate variable
+    next_tick =
```
Comment (Neylix) [inline lib/archethic/contracts/worker.ex:207]: Same here for `Utils.next_date(interval, DateTime.utc_now())`

File: lib/archethic/utils/regression/benchmarks/end_to_end_validation.ex [hunk 3]
```diff
@@ -29,7 +29,15 @@ defmodule Archethic.Utils.Regression.Benchmark.EndToEndValidation do
 
     {:ok, pid} = SeedHolder.start_link(seeds: seeds)
 
-    allocate_funds(SeedHolder.get_seeds(pid), host, port)
+    amount = 100
+
+    Playbook.batch_send_funds_to(
+      SeedHolder.get_seeds(pid)
```
Comment (Neylix) [inline lib/archethic/utils/regression/benchmarks/end_to_end_validation.ex:35]: Here you should pass a list of addresses and not seeds

File: lib/archethic/utils/regression/playbooks/smart_contract.ex [hunk 4]
```diff
@@ -9,94 +9,136 @@ defmodule Archethic.Utils.Regression.Playbook.SmartContract do
...
         },
         host,
         port
       )
 
-    Logger.info(
-      "Deployed smart contract at #{Base.encode16(contract_tx_address)} which sends 0.1 UCO to #{Base.encode16(recipient_address2)} each seconds"
-    )
+    # wait some time for the contract to trigger and create new tx
+    Process.sleep(250)
```
Comment (Neylix) [inline lib/archethic/utils/regression/playbooks/smart_contract.ex:95]: Here you may request the last address of the SC chain and wait for it to change (with a max timeout). You can take inspiration from `Utils.await_confirmation`

File: lib/archethic/utils/regression/playbooks/smart_contract.ex [hunk 5]
```diff
@@ -9,94 +9,136 @@ defmodule Archethic.Utils.Regression.Playbook.SmartContract do
...
+    trigger_address
+  end
 
-    balance =
-      Enum.reduce(0..4, 0, fn _i, _acc ->
-        Process.sleep(1_000)
-        balance = Playbook.get_uco_balance(recipient_address2, host, port)
-        Logger.info("#{Base.encode16(recipient_address2)} received #{balance} UCO")
-        balance
+  def request_funds(amount_by_seed, host, port) do
```
Comment (Neylix) [inline lib/archethic/utils/regression/playbooks/smart_contract.ex:101]: You may move this function to playbook and use it from SC as well as benchmarks

**Comment by bchamagne** [comment]
Depends on https://github.com/archethic-foundation/archethic-node/pull/1124. It cannot work without it.

**Pull request merged**